### PR TITLE
Check whether calculation finished successfully

### PR DESCRIPTION
**Pull request by danielhollas**
CClib provides a field for checking whether calculation finished normally so we should use it.

Note that I am purposely returning the error code at the end, after the output dictionary is assigned to an output node, because it might still contain useful information even from incomplete job.

### CODE REVIEW

**Comment by danielhollas** [comment]
The CI failure for Python 3.10 seems to be unrelated to this PR.

**Comment by danielhollas** [comment]
The CI failure for Python 3.10 is due to a mypy bug for Python 3.10.7, which is fixed by newer mypy version. 
https://github.com/python/mypy/issues/13627

I will fix it as part of #46,

**Comment by danielhollas** [comment]
I've added two new parser tests that cover the new code path so this is ready to go @pzarabadip 

@sphuber in case you have a few minutes, would you mind taking a look as well if the new parser tests make sense to you? Thanks! (btw: already in this PR the existing parser test uncovered a bug I introduced. :smile: )

File: tests/parsers/test_orca.py [hunk 1]
```diff
@@ -26,3 +26,49 @@ def test_orca_default(aiida_localhost, generate_calc_job_node, generate_parser,
...
+    # Pop the cell if it is there, ORCA does not support periodic cell calculations
+    structure_attributes.pop('cell', None)
+
+    data_regression.check({
+        'relaxed_structure': structure_attributes,
+        'output_parameters': results['output_parameters'].attributes,
+    })
+
+
+def test_orca_missing_stdout(aiida_localhost, generate_calc_job_node, generate_parser, generate_inputs_orca):
```
Comment (danielhollas) [inline tests/parsers/test_orca.py:80]: This new test is not strictly related to this PR, but I though it is useful to add to have more coverage of the parser error paths. It checks the following code

https://github.com/pzarabadip/aiida-orca/blob/4a3737a98f8458837ebfaf81d1dd9eb1c41ea953/aiida_orca/parsers/__init__.py#L40

File: tests/parsers/test_orca.py [hunk 2]
```diff
@@ -26,3 +26,49 @@ def test_orca_default(aiida_localhost, generate_calc_job_node, generate_parser,
         'relaxed_structure': structure_attributes,
         'output_parameters': results['output_parameters'].attributes,
     })
+
+
+def test_orca_unsuccessfull(
```
Comment (sphuber) [inline tests/parsers/test_orca.py]:
```suggestion
def test_orca_unsuccessful(
```
But maybe a better name is `failed`? This is also the language used in `aiida-core` for a job with non-zero exit status

File: tests/parsers/fixtures/orca/unsuccessfull/aiida.out [hunk 3]
```diff
@@ -0,0 +1,2244 @@
...
+
+Total enthalpy                    ...    -39.91509734 Eh 
+Total entropy correction          ...     -0.02113973 Eh    -13.27 kcal/mol
+-----------------------------------------------------------------------
+Final Gibbs free energy         ...    -39.93623707 Eh
+
+For completeness - the Gibbs free energy minus the electronic energy
+G-E(el)                           ...      0.03173838 Eh     19.92 kcal/mol
+
+NOTE(danielhollas): The end of this file was purposfully deleted,
```
Comment (sphuber) [inline tests/parsers/fixtures/orca/unsuccessfull/aiida.out]: If the point of this test is to just check that you get `ERROR_CALCULATION_UNSUCCESSFULL` why do you need all this output? Unless you explicitly want to test that you still parse some explicit partial output, you may want to cut a lot more output. This is from the perspective of not bloating the repo too much. As you start adding more tests, if each tests adds thousands of lines, this may start to grow the repo unnecessarily.

File: tests/parsers/test_orca.py [hunk 4]
```diff
@@ -26,3 +26,49 @@ def test_orca_default(aiida_localhost, generate_calc_job_node, generate_parser,
...
+    name = 'unsuccessfull'
+    entry_point_calc_job = 'orca.orca'
+    entry_point_parser = 'orca_base_parser'
+
+    node = generate_calc_job_node(entry_point_calc_job, aiida_localhost, name, generate_inputs_orca())
+    parser = generate_parser(entry_point_parser)
+    results, calcfunction = parser.parse_from_node(node)
+
+    assert calcfunction.is_finished, calcfunction.exception
+    assert calcfunction.exit_status > 299
```
Comment (sphuber) [inline tests/parsers/test_orca.py]: I would advice checking for an explicit exit code. You can use `OrcaCalculation.exit_codes.ERROR_CALCULATION_UNSUCCESSFULL.value`

File: tests/parsers/test_orca.py [hunk 5]
```diff
@@ -26,3 +26,49 @@ def test_orca_default(aiida_localhost, generate_calc_job_node, generate_parser,
...
+    assert not calcfunction.is_finished_ok, calcfunction.exit_message
```
Comment (sphuber) [inline tests/parsers/test_orca.py]: if you check `assert calculation.is_finished` and the `exit_status`, this check is superfluous really

File: tests/parsers/test_orca.py [hunk 6]
```diff
@@ -26,3 +26,49 @@ def test_orca_default(aiida_localhost, generate_calc_job_node, generate_parser,
...
```
Comment (sphuber) [inline tests/parsers/test_orca.py]:
```suggestion
    assert calcfunction.is_finished, calcfunction.exception
    assert calcfunction.exit_status > 299
```
Same here, just check the exact exit code that is expected.

File: aiida_orca/parsers/__init__.py [hunk 7]
```diff
@@ -88,10 +88,12 @@ def _remove_nan(parsed_dictionary: dict) -> dict:
             relaxed_structure = StructureData(ase=ase_structure)
             self.out('relaxed_structure', relaxed_structure)
 
-        pt = PeriodicTable()  # pylint: disable=invalid-name
-
-        output_dict['elements'] = [pt.element[Z] for Z in output_dict['atomnos'].tolist()]
+        if output_dict.get('atomnos') is not None:
```
Comment (sphuber) [inline aiida_orca/parsers/__init__.py:101]:
```suggestion
        if output_dict.get('atomnos', None) is not None:
```
for safety?

**Review by sphuber — CHANGES_REQUESTED** [review]
The tests look good to me in general @danielhollas . Just have some minor suggestions

File: tests/parsers/fixtures/orca/unsuccessfull/aiida.out (hunk 3 above)
Comment (danielhollas) [reply tests/parsers/fixtures/orca/unsuccessfull/aiida.out]: Yes, the point of this was to indeed verify that the output was still parsed in the same way as in the default test.
Agreed about the repo bloat. I'd probably keep this one anyway if @pzarabadip agrees.

File: aiida_orca/parsers/__init__.py (hunk 7 above)
Comment (danielhollas) [reply aiida_orca/parsers/__init__.py:101]: I checked the Python docs and dict.get() should return None by default if the key is missing.

File: tests/parsers/test_orca.py (hunk 5 above)
Comment (danielhollas) [reply tests/parsers/test_orca.py]: Thanks, done.

File: tests/parsers/test_orca.py (hunk 2 above)
Comment (danielhollas) [reply tests/parsers/test_orca.py]: Thanks, I fixed the spelling, but otherwise kept it as is, to distinguish this test from from other possible failure modes in the parser.

File: tests/parsers/test_orca.py (hunk 4 above)
Comment (danielhollas) [reply tests/parsers/test_orca.py]: Cool, I wasn't sure how to get to the exit code, thanks! (Note that ExitCode doesn't have `value` attribute, I used `status` instead.)

File: aiida_orca/parsers/__init__.py (hunk 7 above)
Comment (sphuber) [reply aiida_orca/parsers/__init__.py:101]: Wow, you are absolutely right. I was almost certain this wasn't the case. What the hell 😄

File: tests/parsers/test_orca.py (hunk 4 above)
Comment (sphuber) [reply tests/parsers/test_orca.py]: yeah that's the one

**Review by ezpzbz — APPROVED** [review]
